src/pipeline.cc

```diff
@@ -712,47 +719,86 @@ class PipelineWorker : public AsyncWorker {
-        return Error();
-      }
 
-      // Ensure overlay is sRGB
-      VipsImage *overlayImageRGB;
-      if (vips_colourspace(overlayImage, &overlayImageRGB, VIPS_INTERPRETATION_sRGB, NULL)) {
-        return Error();
+      VipsImage *overlayImageRGB = NULL;
+
+      if (hasOverlayPath) {
+        VipsImage *overlayImage = NULL;
+        ImageType overlayImageType = ImageType::UNKNOWN;
+        overlayImageType = DetermineImageType(baton->overlayPath.c_str());
+        if (overlayImageType != ImageType::UNKNOWN) {
+          overlayImage = InitImage(baton->overlayPath.c_str(), baton->accessMethod);
+          if (overlayImage == NULL) {
+            (baton->err).append("Overlay image has corrupt header");
+            return Error();
+          } else {
+            vips_object_local(hook, overlayImage);
+          }
+        } else {
+          (baton->err).append("Overlay image is of an unsupported image format");
+          return Error();
+        }
+        if (overlayImage->BandFmt != VIPS_FORMAT_UCHAR && overlayImage->BandFmt != VIPS_FORMAT_FLOAT) {
+          (baton->err).append("Expected overlay image band format to be uchar or float: ");
+          (baton->err).append(vips_enum_nick(VIPS_TYPE_BAND_FORMAT, overlayImage->BandFmt));
+          return Error();
+        }
+        if (!HasAlpha(overlayImage)) {
+          (baton->err).append("Overlay image must have an alpha channel");
+          return Error();
+        }
+        if (overlayImage->Xsize != image->Xsize && overlayImage->Ysize != image->Ysize) {
+          (baton->err).append("Overlay image must have same dimensions as resized image");
+          return Error();
+        }
+
+        if (vips_colourspace(overlayImage, &overlayImageRGB, VIPS_INTERPRETATION_sRGB, NULL)) {
+          return Error();
+        }
+        vips_object_local(hook, overlayImageRGB);
+      } else {
+        VipsImage *black = NULL;
+
+        if (vips_black(&black, 1, 1, "bands", 4, NULL)) {
+          (baton->err).append("Unable to apply overlay with color");
+          return Error();
+        }
+        vips_object_local(hook, black);
+
+        VipsImage *overlayColor = NULL;
+        double ones[4] = {1, 1, 1, 1};
+        if (vips_linear(black, &overlayColor, ones, baton->overlayColor, 4, NULL)) {
+          (baton->err).append("Unable to apply overlay with color");
+          return Error();
+        }
+        vips_object_local(hook, overlayColor);
+
+        VipsImage *overlayColorUCHAR = NULL;
+        if (vips_cast(overlayColor, &overlayColorUCHAR, VIPS_FORMAT_UCHAR, NULL)) {
+          (baton->err).append("Unable to apply overlay with color");
+          return Error();
+        }
+        vips_object_local(hook, overlayColorUCHAR);
+
+        VipsImage *overlayColorUCHARsRGB = NULL;
+        if (vips_copy(overlayColorUCHAR, &overlayColorUCHARsRGB, "interpretation", VIPS_INTERPRETATION_sRGB, NULL)) {
```

Comment:
Using `vips_colourspace` might make what's happening here be a bit more self-descriptive, but I'm unsure it'll be any faster/better.